File: src/net/HttpProjectRepository.cpp
#include "HttpProjectRepository.h"

#include "domain/AssessmentStatus.h"
#include "net/HttpJson.h"

#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>

HttpProjectRepository::HttpProjectRepository(ApiClient &client)
    : m_client(client)
{
}

QList<Project> HttpProjectRepository::loadProjects() const
{
    int status = 0;
    const QJsonDocument doc = m_client.get(QStringLiteral("/api/v1/projects"), &status);
    if (status != 200 || !doc.isArray()) {
        m_lastError = m_client.lastError();
        return {};
    }

    QList<Project> projects;
    for (const QJsonValue &value : doc.array()) {
        if (value.isObject())
            projects.append(projectFromJson(value.toObject()));
    }
    return projects;
}

Project HttpProjectRepository::createProject(const QString &name, const QString &description,
                                             const QString &catalogVersion)
{
    QJsonObject body;
    body.insert(QStringLiteral("name"), name);
    body.insert(QStringLiteral("description"), description);
    body.insert(QStringLiteral("catalogVersion"), catalogVersion);

    int status = 0;
    const QJsonDocument doc = m_client.post(QStringLiteral("/api/v1/projects"), body, &status);
    if (status != 201 || !doc.isObject()) {
        m_lastError = m_client.lastError();
        return {};
    }
    return projectFromJson(doc.object());
}

bool HttpProjectRepository::updateProject(const Project &project)
{
    QJsonObject body;
    body.insert(QStringLiteral("name"), project.name);
    body.insert(QStringLiteral("description"), project.description);

    int status = 0;
    m_client.patch(QStringLiteral("/api/v1/projects/%1").arg(project.id), body, &status);
    if (status != 200) {
        m_lastError = m_client.lastError();
        return false;
    }
    return true;
}

bool HttpProjectRepository::deleteProject(int projectId)
{
    int status = 0;
    if (!m_client.del(QStringLiteral("/api/v1/projects/%1").arg(projectId), &status)) {
        m_lastError = m_client.lastError();
        return false;
    }
    return status == 204;
}

RequirementAssessment HttpProjectRepository::loadAssessment(int projectId, int targetObjectId,
                                                          int requirementDbId) const
{
    int status = 0;
    const QJsonDocument doc = m_client.get(
        QStringLiteral("/api/v1/projects/%1/target-objects/%2/requirements/%3/assessment")
            .arg(projectId)
            .arg(targetObjectId)
            .arg(requirementDbId),
        &status);
    if (status != 200 || !doc.isObject()) {
        m_lastError = m_client.lastError();
        return {};
    }
    return assessmentFromJson(doc.object());
}

AssessmentSaveResult HttpProjectRepository::saveAssessment(const RequirementAssessment &assessment)
{
    QJsonObject body;
    body.insert(QStringLiteral("status"), assessmentStatusToString(assessment.status));
    body.insert(QStringLiteral("note"), assessment.note);
    body.insert(QStringLiteral("responsible"), assessment.responsible);
    if (assessment.dueDate.isValid())
        body.insert(QStringLiteral("dueDate"), assessment.dueDate.toString(Qt::ISODate));
    body.insert(QStringLiteral("version"), assessment.version);

    int status = 0;
    const QJsonDocument doc = m_client.put(
        QStringLiteral("/api/v1/projects/%1/target-objects/%2/requirements/%3/assessment")
            .arg(assessment.projectId)
            .arg(assessment.targetObjectId)
            .arg(assessment.requirementDbId),
        body, &status);
    if (status == 200 && doc.isObject())
        return AssessmentSaveResult::ok(assessmentFromJson(doc.object()));

    if (status == 409 && doc.isObject()) {
        const QJsonObject obj = doc.object();
        const QJsonValue current = obj.value(QStringLiteral("current"));
        if (current.isObject())
            return AssessmentSaveResult::conflict(assessmentFromJson(current.toObject()));
    }

    m_lastError = m_client.lastError();
    return AssessmentSaveResult::failed();
}

QString HttpProjectRepository::lastError() const
{
    if (!m_lastError.isEmpty())
        return m_lastError;
    return m_client.lastError();
}

File: Delphi/Net/HttpProjectRepository.pas
unit HttpProjectRepository;

interface

uses
  System.SysUtils, System.JSON, IsmsDomain, RepositoryBase, ApiClient, HttpJson;

type
  THttpProjectRepository = class(TProjectRepositoryBase)
  private
    FClient: TApiClient;
    FLastError: string;
  public
    constructor Create(AClient: TApiClient);
    function LoadProjects: TArray<TProject>; override;
    function CreateProject(const AName, ADescription, ACatalogVersion: string): TProject; override;
    function UpdateProject(const AProject: TProject): Boolean; override;
    function DeleteProject(AProjectId: Integer): Boolean; override;
    function LoadAssessment(AProjectId, ATargetObjectId, ARequirementDbId: Integer): TRequirementAssessment; override;
    function SaveAssessment(const AAssessment: TRequirementAssessment): Boolean; override;
    function GetLastError: string; override;
  end;

implementation

constructor THttpProjectRepository.Create(AClient: TApiClient);
begin
  inherited Create;
  FClient := AClient;
end;

function THttpProjectRepository.LoadProjects: TArray<TProject>;
var
  Doc: TJSONValue;
  Arr: TJSONArray;
  I: Integer;
  List: TArray<TProject>;
  Status: Integer;
begin
  SetLength(List, 0);
  Doc := FClient.Get('/api/v1/projects', Status);
  try
    if (Status <> 200) or not (Doc is TJSONArray) then
    begin
      FLastError := FClient.LastError;
      Exit;
    end;
    Arr := TJSONArray(Doc);
    SetLength(List, Arr.Count);
    for I := 0 to Arr.Count - 1 do
      if Arr.Items[I] is TJSONObject then
        List[I] := ProjectFromJson(TJSONObject(Arr.Items[I]));
  finally
    Doc.Free;
  end;
  Result := List;
end;

function THttpProjectRepository.CreateProject(const AName, ADescription,
  ACatalogVersion: string): TProject;
var
  Body: TJSONObject;
  Doc: TJSONValue;
  Status: Integer;
begin
  FillChar(Result, SizeOf(Result), 0);
  Body := TJSONObject.Create;
  try
    Body.AddPair('name', AName);
    Body.AddPair('description', ADescription);
    Body.AddPair('catalogVersion', ACatalogVersion);
    Doc := FClient.PostJson('/api/v1/projects', Body, Status);
    try
      if (Status <> 201) or not (Doc is TJSONObject) then
      begin
        FLastError := FClient.LastError;
        Exit;
      end;
      Result := ProjectFromJson(TJSONObject(Doc));
    finally
      Doc.Free;
    end;
  finally
    Body.Free;
  end;
end;

function THttpProjectRepository.UpdateProject(const AProject: TProject): Boolean;
var
  Body: TJSONObject;
  Doc: TJSONValue;
  Status: Integer;
begin
  Result := False;
  Body := TJSONObject.Create;
  try
    Body.AddPair('name', AProject.Name);
    Body.AddPair('description', AProject.Description);
    Doc := FClient.PatchJson(Format('/api/v1/projects/%d', [AProject.Id]), Body, Status);
    try
      if Status <> 200 then
        FLastError := FClient.LastError
      else
        Result := True;
    finally
      Doc.Free;
    end;
  finally
    Body.Free;
  end;
end;

function THttpProjectRepository.DeleteProject(AProjectId: Integer): Boolean;
var
  Status: Integer;
begin
  Result := FClient.Delete(Format('/api/v1/projects/%d', [AProjectId]), Status) and (Status = 204);
  if not Result then
    FLastError := FClient.LastError;
end;

function THttpProjectRepository.LoadAssessment(AProjectId, ATargetObjectId,
  ARequirementDbId: Integer): TRequirementAssessment;
var
  Doc: TJSONValue;
  Status: Integer;
begin
  FillChar(Result, SizeOf(Result), 0);
  Result.ProjectId := AProjectId;
  Result.TargetObjectId := ATargetObjectId;
  Result.RequirementDbId := ARequirementDbId;
  Result.Status := asOpen;
  Doc := FClient.Get(
    Format('/api/v1/projects/%d/target-objects/%d/requirements/%d/assessment',
      [AProjectId, ATargetObjectId, ARequirementDbId]), Status);
  try
    if (Status <> 200) or not (Doc is TJSONObject) then
    begin
      FLastError := FClient.LastError;
      Exit;
    end;
    Result := AssessmentFromJson(TJSONObject(Doc));
  finally
    Doc.Free;
  end;
end;

function THttpProjectRepository.SaveAssessment(const AAssessment: TRequirementAssessment): Boolean;
var
  Body: TJSONObject;
  Doc: TJSONValue;
  Status: Integer;
begin
  Result := False;
  Body := TJSONObject.Create;
  try
    Body.AddPair('status', AssessmentStatusToString(AAssessment.Status));
    Body.AddPair('note', AAssessment.Note);
    Body.AddPair('responsible', AAssessment.Responsible);
    if IsValidDate(AAssessment.DueDate) then
      Body.AddPair('dueDate', DateToIso(AAssessment.DueDate));
    Body.AddPair('version', TJSONNumber.Create(AAssessment.Version));
    Doc := FClient.PutJson(
      Format('/api/v1/projects/%d/target-objects/%d/requirements/%d/assessment',
        [AAssessment.ProjectId, AAssessment.TargetObjectId, AAssessment.RequirementDbId]),
      Body, Status);
    try
      if Status = 403 then
      begin
        FLastError := 'Keine Berechtigung zum Speichern.';
        Exit;
      end;
      if Status = 409 then
      begin
        FLastError := 'Datensatz wurde zwischenzeitlich geändert. Bitte neu laden.';
        Exit;
      end;
      if (Status <> 200) or not (Doc is TJSONObject) then
        FLastError := FClient.LastError
      else
        Result := True;
    finally
      Doc.Free;
    end;
  finally
    Body.Free;
  end;
end;

function THttpProjectRepository.GetLastError: string;
begin
  if FLastError <> '' then
    Exit(FLastError);
  Result := FClient.LastError;
end;

end.
